src/TextInputWidget.cc

```diff
@@ -65,6 +108,65 @@ FilteredTextEdit::stopTyping()
         emit stoppedTyping();
 }
 
+QSize
+FilteredTextEdit::sizeHint() const
+{
+        ensurePolished();
+        auto margins = viewportMargins();
+        margins += document()->documentMargin();
+        QSize size = document()->size().toSize();
+        size.rwidth() += margins.left() + margins.right();
+        size.rheight() += margins.top() + margins.bottom();
+        return size;
+}
+
+QSize
+FilteredTextEdit::minimumSizeHint() const
+{
+        ensurePolished();
+        auto margins = viewportMargins();
+        margins += document()->documentMargin();
+        margins += contentsMargins();
+        QSize size(fontMetrics().averageCharWidth() * 10,
+                   fontMetrics().lineSpacing() + margins.top() + margins.bottom());
+        return size;
+}
+
+void
+FilteredTextEdit::submit()
+{
+        if (true_history_.size() == INPUT_HISTORY_SIZE)
+                true_history_.pop_back();
+        true_history_.push_front(toPlainText());
+        working_history_ = true_history_;
```

Comment:
This causes the working history to falsely have 127 (mostly empty) entries. If a message is sent you can keep going up to the working history. 

It can be resolved by initializing `true_history_` to 0.

Reply:
Good catch, thanks! Not sure what I was thinking there.